README.md
# simple_tags

[![pub package](https://img.shields.io/pub/v/simple_tags.svg)](https://pub.dev/packages/simple_tags)

A Flutter plugin for displaying simple string based tags inside a wrap.
The plugin is kept extremely simple and straight forward.

## Usage
To use this plugin, add `simple_tags` as a [dependency in your pubspec.yaml file](https://flutter.dev/platform-plugins/).

```
dependencies:
  simple_tags: "0.0.6"
```

### Example

You can specify the `content` of `SimpleTags` by injecting a list of strings.
This will create a basic tag layout without styling.

The example below shows some styling options like for example an altered `TextStyle` and an altered `BoxDecoration` of the tag container.
You can adapt the tag text as well as the tag container toward your needs.
Some basic callsbacks are also provided for tapping or pressing a tag container.

``` dart
class SimpleTagExample extends StatefulWidget {
  SimpleTagExample({Key key, this.title}) : super(key: key);

  final String title;
  final List<String> content = [
    'Apple',
    'Banana',
    'Orange',
    'Pomme',
    'Carambola',
    'Cherries',
    'Date Fruit',
    'A Fruit with a really long name',
    'A Fruit with a really, really, really, really, very, very, very, very long name',
    'Durian',
    'Grapefruit'
  ];

  @override
  _SimpleTagExampleState createState() => _SimpleTagExampleState();
}

class _SimpleTagExampleState extends State<SimpleTagExample> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SimpleTags(
        content: widget.content,
        wrapSpacing: 4,
        wrapRunSpacing: 4,
        onTagPress: (tag) {print('pressed $tag');},
        onTagLongPress: (tag) {print('long pressed $tag');},
        onTagDoubleTap: (tag) {print('double tapped $tag');},
        tagContainerPadding: EdgeInsets.all(6),
        tagTextStyle: TextStyle(color: Colors.deepPurple),
        tagIcon: Icon(Icons.clear, size: 12),
        tagContainerDecoration: BoxDecoration(
          color: Colors.white,
          border: Border.all(color: Colors.grey),
          borderRadius: BorderRadius.all(
            Radius.circular(20),
          ),
          boxShadow: [
            BoxShadow(
              color: Color.fromRGBO(139, 139, 142, 0.16),
              spreadRadius: 1,
              blurRadius: 1,
              offset: Offset(1.75, 3.5), // c
            )
          ],
        ),
      ),
    );
  }
}
```

![image](https://user-images.githubusercontent.com/12799722/229825332-17ed2ef6-080a-43ec-a481-5528f553e47a.png)

## Contributing
Thanks for your interest! Do not hesitate to open an issue if you have a question, feedback or found something that is not working like it should.

Pull requests for improvements of the library or it's documentation are also highly appreciated.

## Licenses
This library and its content is released under the [MIT License](https://choosealicense.com/licenses/mit/).


CHANGELOG.md
## [0.0.6] - 04.04.2023

* Added missing android exported flag for Android 12 and higher

## [0.0.5] - 04.04.2023

* Fixed a bug where too long tag content lead to an overflow

## [0.0.4] - 12.04.2021

* Null safety
* File formatting

## [0.0.3+1] - 12.04.2021

* Altered readme

## [0.0.3] - 12.04.2021

* Removed author and improved description

## [0.0.2] - 12.04.2021

* Added some additional options for text specification and styling

## [0.0.1] - 12.04.2021

* Initial Release


test/simple_tags_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:simple_tags/simple_tags.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/simple_tags.dart
library simple_tags;

import 'package:flutter/cupertino.dart';
import 'package:simple_tags/tag_container.dart';

class SimpleTags extends StatelessWidget {
  /// List of Strings which are displayed by the package.
  final List<String> content;

  /// Function which gets called by pressing on a tag.
  final Function(String)? onTagPress;

  /// Function which gets called by a double tap on a tag.
  final Function(String)? onTagDoubleTap;

  /// Function which gets called by a long press on a tag.
  final Function(String)? onTagLongPress;

  /// Specifies the decoration for the tag container.
  final BoxDecoration? tagContainerDecoration;

  /// Specifies whether there is a trailing widget inside the tag container.
  /// If this is `null` no widget will be displayed.
  final Widget? tagIcon;

  /// Specifies the padding of the tag container.
  final EdgeInsets tagContainerPadding;

  /// Specifies the margin of the tag container.
  final EdgeInsets tagContainerMargin;

  /// Specifies the Textstyle widget for the tag.
  final TextStyle? tagTextStyle;

  /// Specifies the max lines for the tag text.
  final int? tagTextMaxlines;

  /// Specifies the overflow for the tag text.
  final TextOverflow? tagTextOverflow;

  /// Specifies the text align for the tag text.
  final TextAlign? tagTextAlign;

  /// Specifies the softwrapping for the tag text.
  final bool? tagTextSoftWrap;

  /// Specifies the locale for the tag text.
  final Locale? tagTextLocale;

  /// See `WrapCrossAlignment`
  final WrapCrossAlignment wrapCrossAxisAlignment;

  /// See `WrapAlignment`
  final WrapAlignment wrapAlignment;

  /// See `Wrap`
  final double wrapRunSpacing;

  /// See `Wrap`
  final double wrapSpacing;

  /// See `WrapAlignment`
  final WrapAlignment wrapRunAlignment;

  /// See `Axis`
  final Axis wrapDirection;

  /// See `TextDirection`
  final TextDirection? wrapTextDirection;

  /// See `Clip`
  final Clip wrapClipBehavior;

  /// See `VerticalDirection`
  final VerticalDirection wrapVerticalDirection;

  SimpleTags(
      {Key? key,
      required this.content,
      this.onTagPress,
      this.onTagDoubleTap,
      this.onTagLongPress,
      this.tagContainerDecoration,
      this.tagIcon,
      this.tagContainerPadding = EdgeInsets.zero,
      this.tagContainerMargin = EdgeInsets.zero,
      this.tagTextStyle,
      this.tagTextSoftWrap,
      this.tagTextAlign,
      this.tagTextOverflow,
      this.tagTextMaxlines,
      this.tagTextLocale,
      this.wrapCrossAxisAlignment = WrapCrossAlignment.start,
      this.wrapAlignment = WrapAlignment.start,
      this.wrapRunSpacing = 0,
      this.wrapRunAlignment = WrapAlignment.start,
      this.wrapDirection = Axis.horizontal,
      this.wrapSpacing = 0,
      this.wrapTextDirection,
      this.wrapClipBehavior = Clip.none,
      this.wrapVerticalDirection = VerticalDirection.down})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Wrap(
        crossAxisAlignment: wrapCrossAxisAlignment,
        alignment: wrapAlignment,
        runSpacing: wrapRunSpacing,
        runAlignment: wrapRunAlignment,
        direction: wrapDirection,
        spacing: wrapSpacing,
        textDirection: wrapTextDirection,
        clipBehavior: wrapClipBehavior,
        verticalDirection: wrapVerticalDirection,
        children: _buildTagContent());
  }

  List<Widget> _buildTagContent() {
    List<Widget> toReturn = [];

    if (content.isNotEmpty) {
      content.forEach((tag) {
        print(tag);
        toReturn.add(TagContainer(
          tag: tag,
          tagContainerDecoration: tagContainerDecoration,
          tagContainerMargin: tagContainerMargin,
          tagContainerPadding: tagContainerPadding,
          tagIcon: tagIcon as Icon?,
          tagTextStyle: tagTextStyle,
          tagTextAlign: tagTextAlign,
          tagTextLocale: tagTextLocale,
          tagTextOverflow: tagTextOverflow,
          tagTextMaxLines: tagTextMaxlines,
          tagTextSoftWrap: tagTextSoftWrap,
          onPressed: () {
            if (onTagPress != null) {
              onTagPress!(tag);
            }
          },
          onLongPressed: () {
            if (onTagLongPress != null) {
              onTagLongPress!(tag);
            }
          },
          onDoubleTap: () {
            if (onTagDoubleTap != null) {
              onTagDoubleTap!(tag);
            }
          },
        ));
      });
    }

    return toReturn;
  }
}


lib/tag_container.dart
import 'package:flutter/cupertino.dart';

class TagContainer extends StatefulWidget {
  final String tag;
  final EdgeInsets? tagContainerPadding;
  final EdgeInsets? tagContainerMargin;
  final TextStyle? tagTextStyle;

  final int? tagTextMaxLines;
  final TextOverflow? tagTextOverflow;
  final TextAlign? tagTextAlign;
  final bool? tagTextSoftWrap;
  final Locale? tagTextLocale;

  final GestureTapCallback? onPressed;
  final GestureTapCallback? onLongPressed;
  final GestureTapCallback? onDoubleTap;
  final Icon? tagIcon;
  final BoxDecoration? tagContainerDecoration;

  TagContainer(
      {Key? key,
      required this.tag,
      this.onPressed,
      this.onDoubleTap,
      this.onLongPressed,
      this.tagIcon,
      this.tagContainerMargin,
      this.tagContainerPadding,
      this.tagTextStyle,
      this.tagTextMaxLines,
      this.tagTextOverflow,
      this.tagTextAlign,
      this.tagTextLocale,
      this.tagTextSoftWrap,
      this.tagContainerDecoration})
      : super(key: key);

  @override
  _TagContainerState createState() => _TagContainerState();
}

class _TagContainerState extends State<TagContainer>
    with TickerProviderStateMixin {
  late AnimationController _animationController;
  late Animation<double> _animation;
  static const int _fadeDuration = 450;

  @override
  void initState() {
    super.initState();

    _animationController = AnimationController(
        duration: const Duration(milliseconds: _fadeDuration), vsync: this);
    _animation =
        CurvedAnimation(parent: _animationController, curve: Curves.easeIn);

    _animationController.forward();
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return FadeTransition(
      opacity: _animation,
      child: GestureDetector(
        onTap: widget.onPressed,
        onDoubleTap: widget.onDoubleTap,
        onLongPress: widget.onLongPressed,
        child: Container(
          margin: widget.tagContainerMargin,
          padding: widget.tagContainerPadding,
          decoration: widget.tagContainerDecoration,
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            mainAxisSize: MainAxisSize.min,
            children: [
              Flexible(
                child: Text(
                  widget.tag,
                  style: widget.tagTextStyle,
                  maxLines: widget.tagTextMaxLines,
                  overflow: widget.tagTextOverflow,
                  textAlign: widget.tagTextAlign,
                  softWrap: widget.tagTextSoftWrap,
                  locale: widget.tagTextLocale,
                ),
              ),
              widget.tagIcon != null ? widget.tagIcon! : SizedBox()
            ],
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:simple_tags/simple_tags.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Simple Tags Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: SimpleTagExample(title: 'Simple Tags Demo'),
    );
  }
}

class SimpleTagExample extends StatefulWidget {
  SimpleTagExample({Key key, this.title}) : super(key: key);

  final String title;
  final List<String> content = [
    'Apple',
    'Banana',
    'Orange',
    'Pomme',
    'Carambola',
    'Cherries',
    'Date Fruit',
    'A Fruit with a really long name',
    'A Fruit with a really, really, really, really, very, very, very, very long name',
    'Durian',
    'Grapefruit'
  ];

  @override
  _SimpleTagExampleState createState() => _SimpleTagExampleState();
}

class _SimpleTagExampleState extends State<SimpleTagExample> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SimpleTags(
        content: widget.content,
        wrapSpacing: 4,
        wrapRunSpacing: 4,
        onTagPress: (tag) {print('pressed $tag');},
        onTagLongPress: (tag) {print('long pressed $tag');},
        onTagDoubleTap: (tag) {print('double tapped $tag');},
        tagContainerPadding: EdgeInsets.all(6),
        tagTextStyle: TextStyle(color: Colors.deepPurple),
        tagIcon: Icon(Icons.clear, size: 12),
        tagContainerDecoration: BoxDecoration(
          color: Colors.white,
          border: Border.all(color: Colors.grey),
          borderRadius: BorderRadius.all(
            Radius.circular(20),
          ),
          boxShadow: [
            BoxShadow(
              color: Color.fromRGBO(139, 139, 142, 0.16),
              spreadRadius: 1,
              blurRadius: 1,
              offset: Offset(1.75, 3.5), // c
            )
          ],
        ), // This trailing comma makes auto-formatting nicer for build methods.
      ),
    );
  }
}
